src/dataflow/operators/notificator.rs

```diff
@@ -0,0 +1,136 @@
+    pub fn frontier(&self, input: usize) -> &[T] {
+        self.frontier[input].elements()
+    }
+
+    /// Requests a notification at the time associated with capability `cap`. Takes ownership of
+    /// the capability.
+    ///
+    /// In order to request a notification at future timestamp, obtain a capability for the new
+    /// timestamp first, as show in the example.
+    ///
+    /// #Examples
+    /// ```
+    /// use timely::dataflow::operators::{ToStream, Unary};
+    /// use timely::dataflow::channels::pact::Pipeline;
+    ///
+    /// timely::example(|scope| {
+    ///     (0..10).to_stream(scope)
+    ///            .unary_notify(Pipeline, "example", Vec::new(), |input, output, notificator| {
+    ///                while let Some((cap, data)) = input.next() {
+    ///                    output.session(&cap).give_content(data);
+    ///                    let mut time = cap.time();
+    ///                    time.inner += 1;
+    ///                    notificator.notify_at(cap.delayed(&time));
+    ///                }
+    ///                while let Some((cap, count)) = notificator.next() {
+    ///                    println!("done with time: {:?}", cap.time());
+    ///                }
+    ///            });
+    /// });
+    /// ```
+    #[inline]
+    pub fn notify_at(&mut self, cap: Capability<T>) {
+        self.pending.push(cap);
+    }
+
+    /// Repeatedly calls `logic` till exhaustion of the available notifications.
+    ///
+    /// `logic` receives a capability for `t`, the timestamp being notified and a `count`
+    /// representing how many capabilities were requested for that specific timestamp.
+    #[inline]
+    pub fn for_each<F: FnMut(Capability<T>, i64)>(&mut self, mut logic: F) {
+        while let Some((cap, count)) = self.next() {
+            ::logging::log(&::logging::GUARDED_PROGRESS, true);
+            logic(cap, count);
+            ::logging::log(&::logging::GUARDED_PROGRESS, false);
+        }
+    }
+
+    #[inline]
+    fn scan_for_available_notifications(&mut self) {
+        let mut last = 0;
+        while let Some(position) = (&self.pending[last..]).iter().position(|cap| {
+            let time = cap.time();
+            !self.frontier.iter().any(|x| x.le(&time))
+        }) {
+            let newly_available = self.pending.swap_remove(position + last);
+            let push_new = {
+                let already_available = self.available.iter_mut().find(|&&mut (ref cap, _)| {
+                    cap.time() == newly_available.time()
+                });
+                match already_available {
+                    Some(&mut (_, ref mut count)) => {
+                        *count += 1;
+                        false
+                    },
+                    None => true,
+                }
+            };
+            if push_new {
+                self.available.push_back((newly_available, 1));
+            }
+            last = position;
```

Comment:
I think so, right? `position` is relative to an iterator that starts from `last`, so to slide `last` forward we want `+=`. Let's see what Andrea thinks.